### ISSUE TITLE
Store Custom Roles for user authorities and restrictions in a database instead of enum

### ISSUE THREAD

**Issue by bellaabdelouahab** [ISIL-ESTE/Student-Workflow-Organizer#14]
Storing the authorities and restrictions in a database is more flexible and dynamic.
@muttaqin1

**Comment by muttaqin1** [comment]
we have to store the roles enum too.

**Comment by bellaabdelouahab** [comment]
sure, can you provide the scheam for this

**Comment by muttaqin1** [comment]
sure, i will send a pr soon.
@bellaabdelouahab

**Comment by muttaqin1** [comment]
I have found a issue.
Lets assume our server is deployed. Now we want to create a Action.We have created a Action DELETE_ALL using http://localhost:4000/admin/action/create POST

const Action = {
DELETE_ALL:"DELETE_ALL"
}

The Action we have created now, can we use it?
No! Its useless.  Because we are setting actions to our routes manually =>
Router.get("/.....", restrictTo("ADMIN")(Actions.DELETE_USER)), someController)

We are restricting controller manually.

 my solution is:
 We have to create a route model.
Every route should a have a route  record.
The record will contain :
{
path:"/admin/create",(path will be unique. Easy to find)
Actions:["update_user"]

}

So when we create a action using a POST request. We can assign them to routes by pushing it to the route record's Actions array.
Middlewares will be called in every request. So if we push a Action to a route record.It will start working instantly.and we can add Actions to routes dynamically.

what do you think?
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
I will write you within 4 hours, because it's 5 am now here 😁

**Comment by muttaqin1** [comment]
> I will write you within 4 hours, because it's 5 am now here 😁

11.30 AM here😆

**Comment by bellaabdelouahab** [comment]
Your proposed solution to create a route model and associate actions with routes dynamically sounds like a reasonable approach.

 btw, I have a question.

After giving it some thought, I believe that the authorities and restrictions should be static. This means we don't need to store them in the database. On the other hand, we should keep the roles dynamic because we might need to create custom roles. For example, we could have a role named "restricted_to_b_class" that allows certain actions only for class B.

In summary, we can have custom roles where each role has actions. The actions can have one of four values: update, delete, get, getAll, or create. each of those value assosiated with a model here is a code sample :
`
// User model
// .....
roles: {
  type: Role,
},
// .....


// Role model
// .....
update: {
  type: String, // Changed to single string type instead of array
  enum: ['user', 'project', 'calendar']
},
getAll: {
  type: String, // Changed to single string type instead of array
  enum: ['user', 'project', 'calendar']
}
// .....

`
 I'm unsure whether we should keep both authorities and restrictions or just leave restrictions and remove authorities.

Each new user can start with a user role, and then we can add restrictions. If there are too many restrictions, we can create a new role.

I hope I have explained it clearly. Let me know if you have any further questions.

**Comment by muttaqin1** [comment]
> Your proposed solution to create a route model and associate actions with routes dynamically sounds like a reasonable approach.
>  
>  btw, I have a question. 
> 
> After giving it some thought, I believe that the authorities and restrictions should be static. This means we don't need to store them in the database. On the other hand, we should keep the roles dynamic because we might need to create custom roles. For example, we could have a role named "restricted_to_b_class" that allows certain actions only for class B.
> 
> In summary, we can have custom roles where each role has actions. The actions can have one of four values: update, delete, get, getAll, or create. each of those value assosiated with a model here is a code sample :
> ```js
> // User model
> // .....
> roles: {
>   type: Role,
> },
> // .....
> 
> 
> // Role model
> // .....
> update: {
>   type: [String], // store the models names 
>   enum: ['user', 'project', 'calendar']
> },
> getAll: {
>   type: [String], 
>   enum: ['user', 'project', 'calendar']
> }
> // .....
> 
> ```
>  I'm unsure whether we should keep both authorities and restrictions or just leave restrictions and remove authorities.
> 
> Each new user can start with a user role, and then we can add restrictions. If there are too many restrictions, we can create a new role.
> 
> I hope I have explained it clearly. Let me know if you have any further questions.

You're absolutely right.we can create a role model ,but If we keep ACTIONS immutable then we can achieve all of those things we need without making a huge change on our current feature. 
If our ACTIONS is immutable that means no new actions will exists after the deployment.So our routes which will be protected with restrictTo middleware with some ACTIONS, those ACTIONS won't change.

your sample model code:
```js
// Role model
 // ....
create:{
type:[String]
enum:['user', 'project', 'calander']
},
 update: {
  type: [String], // store the models names 
enum: ['user', 'project', 'calendar']
},
getAll: {
  type: [String], 
  enum: ['user', 'project', 'calendar']
}
 // .....
```
this model is equivalent to this:
```js
const Actions = {
CREATE_USER:"CREATE_USER",
CREATE_PROJECT:"CREATE_PROJECT",
GET_ALL_USER:"GET_ALL_USER",
GET_ALL_PROJECT:"GET_ALL_PROJECT",
//...
}
```
And we can make our current ROLES enum dynamic using a model:
```js
{
name:{
type:String,
enum:"SUPER_ADMIN"
},
authorities:{
type:[String],
enum: Object.values(Actions)
},
restrictions:{
type:[String],
enum: Object.values(Actions)
}
} 
```
Then we can create  some routes to create roles using the actions(constant).The functionality to assign roles to user is already implemented.

I think we dont need the restrictions array.Beacuse we can restrict user by removing one or more ACTIONS from the user authorities array.

what do you think about this? Btw, i am ready to implement your  solution.@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
Issue resolved by PR [#18 ]
